ratpoison.1
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2 of the License, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU General Public License along with
.\" this program; if not, write to the Free Software Foundation, Inc., 59 Temple
.\" Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.\" This manpage is written using the mdoc macro language.
.\" Examples of formatters which support mdoc are groff[1] and mandoc[2].
.\" An mdoc language reference is available[3].
.\"
.\" [1] https://gnu.org/software/groff/
.\" [2] http://mdocml.bsd.lv/
.\" [3] http://mdocml.bsd.lv/mdoc.7.html
.\"
.Dd Aug 22, 2017
.Dt RATPOISON 1
.Os
.Sh NAME
.Nm ratpoison
.Nd window manager without mouse dependency
.Sh SYNOPSIS
.Nm
.Op Fl hv
.Nm
.Op Fl d Ar dpy
.Op Fl s Ar num
.Op Fl f Ar file
.Nm
.Op Fl d Ar dpy
.Op Fl s Ar num
.Op Fl i
.Fl c Ar command Op Fl c Ar command ...
.Sh DESCRIPTION
.Nm
is a Window Manager without fat library dependencies, fancy graphics or
rat dependence.
.Pp
The screen can be split into non-overlapping frames.
All windows are kept maximized inside their frames.
.Pp
All interaction with the window manager is done through
keystrokes.
.Nm
has a prefix map to minimize key clobbering.
.Pp
The options are as follows:
.Bl -tag -width Bs
.It Fl c , Fl \-command
Send ratpoison a command.
There must be a ratpoison instance running as window manager for the
given display/screen for this to work.
Do not forget to quote the command if it contains spaces.
For example:
.Dl Nm Fl c Qq Ar "echo hello world"
.It Fl d , Fl \-display Ar display
Set the X display to use or send commands to.
.It Fl f , Fl \-file Ar filename
Specify an alternate configuration file.
If this is not given,
.Nm
will try
.Pa $HOME/.ratpoisonrc
and if that does not exist
.Pa /etc/ratpoisonrc
and execute each command when starting up.
.It Fl h , Fl \-help
Show summary of options.
.It Fl i , Fl \-interactive
Execute commands given with
.Fl c
or
.Fl \-command
in interactive mode.
That means it will behave exactly as if called with
.Ic C\-t \&:
like prompting for missing arguments and things like that.
.It Fl s , Fl \-screen Ar number
Only use the specified screen.
.It Fl v , Fl \-version
Show version of program.
.El
.Sh KEY BINDINGS
To avoid conflicts with other programs, all default ratpoison key
bindings start with an escape key which, by default, is
.Ic C\-t
(Control\-t).
Some important default key bindings:
.Bl -tag -width Ds
.It Ic C\-t \&?
Show key bindings
.It Ic C\-t c
Start an X terminal
.It Ic C\-t n
Switch to next window
.It Ic C\-t p
Switch to previous window
.It Ic C\-t 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
Switch to window number 1 | 2 | ...
.It Ic C\-t k
Close the current window
.It Ic C\-t K
XKill the current application
.It Ic C\-t s | S
Split the current frame into two vertical | horizontal ones
.It Ic C\-t Tab | Left | Up | Right | Down
Switch to the next | left | top | right | bottom frame.
.It Ic C\-t Q
Make the current frame the only one
.It Ic C\-t \&:
Execute a ratpoison command
.El
.Pp
Further default key bindings can be found in parentheses after the
commands in the next section.
.Sh COMMANDS AND DEFAULT ALIASES
.Bl -tag -width Ds
.It Ic abort
.Pq Ic C\-t C\-g
Do nothing and that successfully.
.Po
Useful if you pressed
.Ic C\-t
in error
.Pc .
.It Ic addhook Ar event command
Add a hook: run
.Ar command
whenever
.Ar event
is called.
Possible events are:
.Bl -tag -width Ds
.It Cm deletewindow
Run after a window is withdrawn.
.It Cm newwindow
Run after a new window is mapped.
.It Cm key
Run whenever a top level key is pressed (by default
.Ic C\-t ) .
.It Cm quit
Run before exiting
.Nm .
.It Cm restart
Run before restarting
.Nm .
.It Cm switchframe
Run after a frame actually switched, but before the window in it is
focused.
.It Cm switchgroup
Run after selecting a new group.
.It Cm switchscreen
Run when the user switches to a different screen.
.It Cm switchwin
Run after a new window is selected.
(With dedication, it may already be inactive again, if it was put into
another frame)
.El
.It Ic alias Ar alias command
Add
.Ar alias
as new way to call
.Ar command .
.It Ic bind Ar key command
alias for
.Qq Ic definekey Li root Ar key command
.It Ic banish
.Pq Ic C\-t b
Banish the rat cursor to the lower right corner of the screen.
.It Ic banishrel
Banish the rat cursor to the lower right corner of the current window.
If there isn't a window in the current frame, it banishes the rat cursor
to the lower right corner of the frame.
.It Ic chdir Op Ar directory
If the optional argument is given, change the current directory
of
.Nm
to
.Ar directory .
If nothing is given, change it to the value of the environment variable
.Qq Ev HOME .
.It Ic clrunmanaged
Clears the unmanaged window list.
.It Ic cnext
Like
.Ic next
but switch to the next window with another resource class than the
current one.
(That means the next window belonging to another type of application
than the current one.)
.It Ic cprev
Like
.Ic prev
but switch to the previous window with another resource class than the
current one.
(That means the previous window belonging to another type of application
than the current one.)
.It Ic colon Ar ratpoison\-command Pq Ic C\-t \&:
Execute
.Ar ratpoison\-command
interactively.
(i.e. ask for possible missing arguments.)
.It Ic cother
Like
.Ic other
but switch to the window of the current group that was last accessed and
has another resource class but is not currently visible.
.It Ic curframe Pq Ic C\-t F
Show a bar marking the current frame.
.It Ic definekey Ar keymap key command
Add a new key binding in
.Ar keymap
for
.Ar key
to execute
.Ar command .
Default keymaps are
.Li top
normally only containing
.Ic C\-t ,
which reads a key from
.Li root ,
containing all the normal commands.
.Pp
Note that you have to describe ":" by "colon", "!" by "exclam" and so on.
If you cannot guess a name of a key, try either
.Ic C\-t Ar key
and look at the error message, or try
.Ic :describekey Li root
and pressing the key.
.It Ic dedicate Op Cm 0 | 1
Consider the current frame dedicated/chaste
.Pq Cm 1
or promiscuous
.Pq Cm 0 .
.Pp
A dedicated frame will not accept new windows.
When new windows are to be focused, they will be opened in a non-dedicated
frame instead.
.Pp
If no argument is given, toggle the current dedicateness.
By default no windows are dedicated.
.It Ic delete Pq Ic C\-t k
Close the current window.
.It Ic delkmap Ar keymap
Deletes the keymap named
.Ar keymap ,
that was generated with
.Ic newkmap .
The keymaps
.Li top
(or whatever was specified by
.Ic set Ar topkmap )
and
.Li root
cannot be deleted.
.It Ic describekey Ar keymap
Grab the next key.
Similar to
.Ic readkey ,
.Ic describekey
shows only the command in
.Ar keymap ,
that would be executed by
.Ic readkey .
.It Ic echo Ar text
Show
.Ar text
as
.Nm
message.
.It Ic escape Ar key
Update the default escape key to
.Ar key .
.Pp
Strictly speaking it updates the
.Ic readkey Li root
command in the keymap
.Li top
to
.Ar key ,
the
.Ic other
binding
in
.Li root
to
.Ar key ,
and
.Ic meta
binding in
.Li root
to
.Ar key
without modifiers or
.Ic "C\-" Ns Ar key
if
.Ar key
has no modifiers.
(If
.Ic set Ar topkmap
was called with an argument other than
.Ar top
that will be used instead of
.Ar top . )
.It Ic exchangedown Pq Ic C\-t C\-Down
Exchange the window in the current frame with the window in the frame
below the current frame.
.It Ic exchangeleft Pq Ic C\-t C\-Left
Exchange the window in the current frame with the window in the frame
left of the current frame.
.It Ic exchangeright Pq Ic C\-t C\-Right
Exchange the window in the current frame with the window in the frame
right of the current frame.
.It Ic exchangeup Pq Ic C\-t C\-Up
Exchange the window in the current frame with the window in the frame
above the current frame.
.It Ic exec Ar shell\-command Pq Ic C\-t \&!
Spawn a shell executing
.Ar shell\-command .
.It Ic execa Ar shell\-command
Spawn a shell executing
.Ar shell\-command ,
without remembering the current frame, so that _NET_WM_PID declaring
programs will be placed into the frame active when they open a window
instead of the frame active when
.Nm
gets this command.
.It Ic execf Ar frame shell\-command
Spawn a shell executing
.Ar shell\-command ,
showing _NET_WM_PID supporting programs in the given frame instead of
the frame selected when this program is run.
.It Ic fdump Op Ar screenno
Output the defining data for all frames of the current screen, or
for screen number
.Ar screenno
if this is specified.
.It Ic focus Pq Ic C\-t Tab
Focus the next frame.
.It Ic focuslast
Switch to the last selected focus.
.It Ic focusleft Pq Ic C\-t Left
Switch to the frame to the left of the current one.
.It Ic focusdown Pq Ic C\-t Down
Switch to the frame beneath the current one.
.It Ic focusright Pq Ic C\-t Right
Switch to the frame to the right of the current one.
.It Ic focusprev
Focus the previous frame.
.It Ic focusup Pq Ic C\-t Up
Switch to the frame above the current one.
.It Ic frestore Ar frames
Replace the current frames with the ones specified in
.Ar frames
in the format as generated by
.Ic fdump .
.It Ic fselect Oo Ar frameno Oc Pq Ic C\-t f
If an argument is supplied, switch to a frame given by number
.Ar frameno .
.Pp
If no argument is given, show a frame selector in each frame and wait for
a key to be pressed.
If the key matches an existing frame selector, this frame gets focused.
.Pp
Frame selectors are by default the numbers starting with zero, but they
can be changed by
.Ic set Ns
ting
.Ar framesels .
.It Ic gdelete Op Ar group
If the optional argument
.Ar group
is supplied, delete
.Ar group .
Otherwise delete the current group.
If the last group is deleted, a new group with name
.Li default
is created.
The group has to be empty, otherwise it cannot be deleted.
.It Ic getenv Ar variable
Output the value of the environment variable
.Ar variable .
.It Ic getsel
Paste the current X Selection into the current window.
.It Ic gmerge Ar group
Move all windows from group
.Ar group
into the current group.
.It Ic gmove Ar group
Move the current window into group
.Ar group .
.It Ic gnew Ar group
Create a new group with name
.Ar group
and select it.
Most window commands only see (and thus select, consider next,
previous or last) windows within the group active when they are
issued.
.It Ic gnewbg Ar group
Create a new group named
.Ar group ,
but do not select it.
.It Ic gnext
Select the next group.
Most window commands only see windows in the effective group.
.It Ic gnumber Op Ar old new
Give the number
.Ar new
to the group with the number
.Ar old
or the current group.
.It Ic gother
Select the last accessed group.
Most window commands only see windows in the effective group.
.It Ic gprev
Select the prior group.
Most window commands only see windows in the effective group.
.It Ic gravity Op Cm nw | w | sw | n | c | s | ne | e | se
Change how in its frame the current window is aligned.
.It Ic grename
Rename current group.
.It Ic groups
Output a list of all groups with their number.
.It Ic gselect Ar group
Select the group named
.Ar group .
.It Ic help Op Ar keymap
If the optional parameter
.Ar keymap
is given, list all keybindings in this keymap, otherwise list all key
bindings in keymap
.Li root .
.It Ic hsplit Oo Ar l Ns Li / Ns Ar p | Oo Li \- Oc Ns Ar pixels Oc Pq Ic C\-t S
Split the current frame into left frame and a right frame.
If no parameter is given, split in halves.
If two numbers separated by a slash
.Pq Ql "/"
are given, the left one is
.Ar l
times the
.Ar p Ns
th
part and the right one
.Pq Ar p Li \- Ar l
times the
.Ar p Ns
th
part of the prior width.
Otherwise the right half is
.Ar pixels
wide or the left one is
.Ar pixels
wide, depending whether there is
.Ql \-
in front of the number or not.
.It Ic inext
Like
.Ic next
but switch to the next window with the same resource class as the
current one.
(That means the next window belonging to the same application
as the current one.)
.It Ic info Pq Ic C\-t i
Output the current the width, height, window number and window name of
the current window.
.Po
What name means is chosen by
.Dq Ic set Ar winname .
.Pc
.It Ic iprev
Like
.Ic prev
but switch to the previous window with the same resource class as the
current one.
(That means the previous window belonging to the same application as the
current one.)
.It Ic iother
Like
.Ic other
but switch to the window of the current group that was last accessed and
has the same resource class but is not currently visible.
.It Ic kill Pq Ic C\-t K
Close the X-connection of the X-client responsible for the current window.
.It Ic lastmsg Pq Ic C\-t m
Reshow the last message.
.It Ic link Ar key Op Ar keymap
Do what
.Ar key
is bound to in the keymap
.Ar keymap
if supplied.
Otherwise what
.Ar key
is bound to in keymap
.Li root .
.It Ic listhook Ar event
List all commands specified with
.Ic addhook
to be executed when even
.Ar event
occurs.
.It Ic meta Oo Ar key Oc Pq Ic C\-t t
Send the escape key (that which normally is
.Ic C\-t )
to the current window.
If a
.Ar key
is specified, this is sent instead.
Note that some applications by default ignore the synthetic key that is
sent using this command as it is considered a security hole.
xterm is one such application.
.It Ic newkmap Ar keymap
Generate a new keymap named
.Ar keymap .
This keymap can be used to add new key-command mappings to it with
.Ic definekey
and can be called with
.Ic readkey .
.It Ic next Pq Ic C\-t Return | C\-t n | C\-t space
Switch to the next window in the current group.
.It Ic nextscreen Pq Ic C\-t N
Switch to the next screen. (If you have multiple physical ones.)
.It Ic number Ar new Op Ar old
Give the number
.Ar new
to the window with the number
.Ar old
or the current window.
.It Ic only Pq Ic C\-t Q
Remove all frames on the current screen except the current frame and
maximize this one to the size of the whole screen.
.It Ic other Pq Ic C\-t C\-t
Switch to the window of the current group that was last
accessed but is not currently visible.
.It Ic prev Pq Ic C\-t p
Switch to the previous window in the current group.
.It Ic prevscreen Pq Ic C\-t P
Switch to the previous screen. (If you have multiple physical ones.)
.It Ic prompt Op Ar prompt
.Nm
will ask the user for input, showing
.Ar prompt
(or a single colon, if no argument is given) and output the input the
user has made.
Note that this command probably does not make much sense in interactive
mode.
.It Ic putsel Ar x\-selection
Replace the X selection with the text
.Ar x\-selection .
It can be inserted into the current window with
.Ic getsel .
.It Ic quit
Quit
.Nm .
.It Ic ratinfo
Display the x y coordinates of the rat cursor relative to the screen.
.It Ic ratrelinfo
Display the x y coordinates of the rat cursor relative to the current
window or current frame if no window is focused
.It Ic ratwarp Ar x y
Move the rat cursor to the position
.Ar ( x , y ) .
.It Ic ratrelwarp Ar deltax deltay
Move the rat cursor to
.Ar ( deltax , deltay ) ,
relative to the current position.
.It Ic ratclick Op Ar button
Simulate a rat click with
.Ar button
(button 1=left button if none given).
.It Ic rathold Cm ( up | down ) Op Ar button
Simulate pressing|releasing rat button
.Ar button
(1=left button if none given).
.It Ic readkey Ar keymap
Grab the next key pressed, and execute the command associated to this key
in
.Ar keymap .
To show it is waiting for a key,
.Nm
will change the rat cursor to a square if
.Va waitcursor
is set.
This command is perhaps best described with its usage in the default
configuration: by pressing
.Ic C\-t ,
which is the only key in the keymap
top ,
the command
.Qq Ic readkey Ar root
is executed.
The next key then executes the command in keymap
.Li root
belonging to
this command.
.It Ic redisplay Pq Ic C\-t l
Extend the current window to the whole size of its current frame and
redisplay it.
(Useful to redisplay normal windows or bring transient windows to the
full size of the frame as only normal windows are maximized by
.Nm )
.It Ic redo Pq Ic C\-t U
Revert the last
.Ic undo
of frame changes.
.It Ic remhook Ar event command
Remove command
.Ar command
from the list of commands to be called when event
.Ar event
is hit.
(The command has to specified, as an event can have multiple commands
attached to it.)
Use
.Qq Ic listhook Ar hook
to get a list of all attached commands.
.It Ic remove Pq Ic C\-t R
Remove the current frame and extend some frames around to fill the
remaining gap.
.It Ic resize Oo Ar deltax deltay Oc Pq Ic C\-t r
If
.Ar deltax
and
.Ar deltay
are supplied, resize the current frame by that (i.e. move the bottom
right corner by the given offsets and then move this frame and resize
adjacent frames to make the frames fill the whole screen again.)
.Pp
If in interactive mode no arguments are supplied, resize the current
frame interactively:
.Pp
.Bl -tag -offset 2n -width "C-f, Right, l" -compact
.It Ic Return
finish resizing
.It Ic C\-g , Escape
abort resizing
.It Ic C\-n , Down , j
grow vertically
.It Ic C\-p , Up , k
shrink vertically
.It Ic C\-f , Right , l
grow horizontally
.It Ic C\-b , Left , h
shrink horizontally
.It Ic s
shrink to size of current window
.El
.Pp
While resizing interactively, changes are in multiples of the amount
of pixels given by
.Ic set Cm resizeunit
(by default 10).
.It Ic restart
Restart
.Nm .
.It Ic sdump
Output the list of all screens.
The screens are separated by commas.
Each screen is shown as 6 values: its number, its x-coordinate, its
y-coordinate, its width, its height and if it is currently selected
(1=true, 0=false).
.It Ic select ( Cm \- | Ar name | Ar number ) Pq Ic C\-t \&'
If a number is given, switch to the window with number
.Ar number .
If a name is given, switch to the window in the current group with
name
.Ar name .
Blank the current frame, if
.Cm \-
is given.
.It Ic set Op Ar variable Op Ar value
If no argument is given, output all
.Nm
variables and their values.
.Pp
If one argument is given, output the value of
.Nm
variable
.Ar variable .
Otherwise set
.Ar variable
to
.Ar value .
What values are valid depends on the variable.
See the section
.Sx VARIABLES
later in this document for details.
.It Ic setenv Ar variable value
Set the environment variable
.Ar variable
to
.Ar value .
.Po
Environment variables will be passed to all programs started from
.Nm .
.Pc
.It Ic sfdump
Output all frames similar to
.Ic fdump ,
but not limited to one screen, but all screens at once and with the
screen number after each frame.
.It Ic sfrestore Ar frames
Replace the current frames with the ones specified in
.Ar frames
in the format as generated by
.Ic sfdump .
.It Ic shrink
Shrink the current frame to the size of the current window with in.
.It Ic split Oo Ar split Oc Pq Ic C\-t s
alias for
.Ic vsplit
.It Ic source Ar file
Read
.Ar file
and execute each line as
.Nm
command.
.It Ic sselect Ar screennumber
Switch to the screen
.Ar screennumber .
(If you have multiple physical ones.)
.It Ic swap Ar dest-frame Oo Ar src-frame Oc Pq Ic C\-t x
Exchange the window in
.Ar src\-frame
(or the current frame if there is no second argument) with the window
.Ar dest\-frame
(or ask interactively which frame to swap with if there is no argument).
.It Ic title Ar newname Pq Ic C\-t A
Overwrite the title of the current window with
.Ar newname .
All following
.Nm
commands will only know the window under the new name.
.It Ic unalias Ar alias
Remove the alias
.Ar alias .
.It Ic unbind Ar key
alias for
.Dl Ic undefinekey Ar root key
.It Ic undefinekey Ar keymap key
Remove the binding for
.Ar key
from
.Ar keymap .
.It Ic undo Pq Ic C\-t _ , C\-t u
Un\-do the last change to the frameset.
(Like splitting, resizing, deleting, ...)
.Pp
The amount of steps that can be undone is specified by the variable
.Va maxundos .
.It Ic unmanage Op Ar name
Add
.Ar name
to the list of unmanaged windows.
Thus, windows of this name will not be managed but allowed to choose
their position themselves.
.Pp
In non\-interactive mode calling it without arguments will print the list.
.Pp
The list can be cleared again by calling
.Ic clrunmanaged .
.It Ic unsetenv Ar variable
Remove variable
.Ar variable
from the list of environment variables.
.It Ic verbexec Ar cmdline
Spawn a shell executing
.Ar cmdline
after showing a message with the command.
.It Ic version Pq Ic C\-t v
Output version and compile time information.
.It Ic vselect Oo Ar vscreenno Oc
Switch to the virtual screen numbered
.Ar vscreenno .
.It Ic vmove Oo Ar vscreenno Oc
Move the current window to the virtual screen numbered
.Ar vscreenno .
.It Ic vsplit Oo Ar l Ns Li / Ns Ar p | Ar "pixels-from-top" | Li \- Ns Ar "pixels-from-bottom" Oc Pq Ic C\-t s
Split the current frame into upper frame and a lower frame.
If no parameter is given, split in halves.
If two numbers separated by a slash
.Pq Dq Li /
are given, the upper one is
.Ar l
times the
.Ar p Ns
th part and the lower one
.Pq Ar p Li \- Ar l
times the
.Ar p Ns
th
part of the prior height.
Otherwise the lower one is
.Ar "pixels from bottom"
wide or the upper one
.Ar "pixels from top"
high, depending whether there is a
.Dq Li \-
in front of the number or not.
.It Ic windows Oo Ar format Oc Pq Ic C\-t w
In interactive mode,
show the list of all windows in the current group for the duration
specified by the variable
.Va msgwait .
If
.Va msgwait
was zero, toggle between indefinitely showing and not showing.
.Pp
The messages are shown in columns or rows depending on the value of
.Va winliststyle
in the format set by
.Ic set Ar winfmt .
The following substitutions happen in format:
.Pp
.Bl -tag -offset 2n -width 2n -compact
.It Li %a
application name (resource name)
.It Li %c
resource class
.It Li %f
frame number
.It Li %g
gravity of the window
.It Li %h
height of the window
.It Li %H
unit to resize the window vertically (height_inc)
.It Li %i
X Window ID
.It Li %p
process ID
.It Li %l
last access number
.It Li %M
string
.Li Maxsize ,
if it specifies a maximum size
.It Li %n
window number
.It Li %s
window status
.Po
.Ql *
is active window,
.Ql +
would be chosen by
.Ic other ,
.Ql \-
otherwise
.Pc
.It Li %S
screen number
.It Li %t
window name
.Po see
.Ic set Ar winname
.Pc ,
.It Li \&%T
the string
.Dq Li Transient ,
if it is a transient window
.It Li %w
width of the window
.It Li %W
unit to resize the window horizontally (width_inc)
.It Li %x
xrandr screen number
.It Li %%
litteral
.Ql %
.El
.Pp
Additionally there can be a positive decimal integer number between the
percent sign and the format string to specify the length this value
should be truncated to if longer.
.Po
For example:
.Li %20t
.Pc
.Pp
In non\-interactive mode, output the list of windows in the current group
line by line.
The format string can be overwritten by the optional parameter
.Ar format .
.El
.Sh VARIABLES
.Nm
variables can be shown and set with
.Ic set .
The following variables are supported:
.Bl -tag -width Ds
.It Cm barborder Ar pixels
Selects how thick the frame around
.Nm Ap
s

NEWS
ratpoison NEWS --- history of user-visible changes.  -*- outline -*-

Copyright (C) 2000, 2001, 2002, 2003, 2004 Shawn Betts
Copyright (C) 2013, 2014, 2017 Jérémie Courrèges-Anglas

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.

* Changes since 1.4.9

** Feature/misc changes

*** manpage fixes

*** fix how find_frame_left, find_frame_right check for overlap

*** always show the frame indicator when switching to a different screen

*** fix bogus libhistory usage

*** fix for tmpwm and multiple screens

*** new 'onlyborder' variable
To print borders only when >=2 frames are on screen.

*** %42s-style format strings now account for UTF-8 characters

** Packaging changes

*** the debian/ subdirectory was removed (outdated)

*** a basic regress test has been added

* Changes since 1.4.8

** Incompatible changes

*** command `compat' and the various def* compat commands were removed

*** `msgwait', `rudeness', `startupmessage', `warp' are now variables
The msgwait, rudeness, startup_message and warp commands will stay
around for a few releases.  Upgrade your config file now!

** Feature/misc changes

*** XRandR support
ratpoison now uses XRandR to detect screens additions/removals at
runtime.  The Xinerama support has been removed.

*** the focus* commands can be used to navigate across screens

*** window completion match by substring, case-insensitive

*** new variable `framemsgwait'
Timeout (in seconds) for the `Current frame' message window.
If set to -1, no message is shown.

*** better `select' error handling

*** the command `set' with no argument dumps a list of all variables/values

*** unknown commands are now logged as warnings to stdout too
Should help debugging a config file.

*** various cleanups and minor bugfixes

** Packaging changes

*** ratpoison now depends on libXrandr instead of libXinerama

*** ratpoison doesn't link against libhistory by default any more
Use ./configure --enable-history to force the use of said library.

*** no ChangeLog file
Please use the git repository instead.  Savannah offers a web interface.

*** no more timestamps in compiled code
Should help reproducible builds.

* Changes since 1.4.7
This is mostly a bugfix release.

** fix cmd_sfdump with multiple screens
Broken since last release, from Joshua Leners.

** respect --enable-history=no
From Jeroen Roovers (jer@gentoo).

** fix for sloppy.c
The ugly sloppy.c had "a bug".  It may be fixed now.

** README improvements
From Jeff Abrahamson.

* Changes since 1.4.6
** More fixes and code cleanup.

** UTF-8 editing support in the input bar

** get rid of "XGetWMName failed" spam on stderr
Introduced in the previous release, sorry!

** groups can now be renumbered (`gnumber')

** contrib/rpdate.sh, fancy calendar shell script to use instead of `time'

** fix `meta', which was sending garbage and made firefox unusable

** handle errors about missing config files
- ratpoison -f missing-config-file -> error
- unreadable ~/.ratpoisonrc file -> warning

** respect --sysconfdir for the global ratpoisonrc file
Rather than a hardcoded /etc/ratpoisonrc location.

** rpws gains a new `current' command (prints the current workspace)

** new manpage, mdoc(7) formatted

** X socket closed before spawning processes
Let's save those precious file descriptors.

** DWIM windows/groups names matching
Window and group names are first looked up for exact match; in last
resort a fuzzy match is attempted.

** try to cope with unset/empty HOME env. variable

** ratpoison.el is now installed at the proper location
ie. in $prefix/share/emacs/site-lisp/

** ratpoison now restricts permissions on your history file

* Changes since 1.4.5
** *Lots* of reliability fixes in ratpoison itself
Please check out the ChangeLog for more details.

** Build process cleanup
e.g. detection of getline() was fixed, build is ok with old gcc's (2.95)

** ratpoison -c now reports errors ($? != 0)

** UTF-8 text is now printed correctly
Both with and without Xft support.

** contrib/ cleanup
Scripts should be more portable, ruby bindings work with newer ruby versions.

** New newwindow and titlechanged hooks

** Functionality was added to rpws
It can now move windows to other workspaces; workspaces can be dumped and
restored.

* Changes since 1.4.4
** bug fixes to history and xft

* Changes since 1.4.3
** add commands ratinfo, ratrelinfo, banishrel

** add history to input bar when libhistory is not used

** add settings historysize and historycompaction

** new window formatters %p, %M
prints the pid of the window and whether its a maxsize window.

** new commands grename, gother

** new bindings C-t N and C-t P
bound to nextscreen/prevscreen

** new settings, bwcolor and fwcolor
For setting the border color of focused and unfocused windows.

** netwm dialog type windows supported
These windows are treated as transients. Some Firefox popup windows
are tagged as dialog windows and should now work with ratpoison as
expected.

** new format character %M
This expands to Maxsize if the window is a maxsize window, nothing if not.

** ratpoison can use xft
If xft is found on your system ratpoison will use it unless explicitely
told not to using --without-xft.

* Changes since 1.4.2
** key event handling
top level key bindings are grabbed "asynchronously" instead of
"synchronously." This should eliminate the freezes some people have
experienced in previous versions.

* Changes since 1.4.1
Just bug fixes.

* Changes since 1.4.0
** set padding affects bar location
On OSX the bar and top of each window is hidden by the menu bar. set
padding corrects the window problem but not the bar problem. So this
change fixes that.

** new set variable framefmt
See docs for info.

** new commands swap, exchangeleft, exchangeright, exchangeup, exchangedown
See docs for info

** new commands execa and execf
See docs for info.

** default font change

contrib/ratpoison.el
;;; ratpoison.el --- ratpoison support for Emacs
;;
;; Copyright (C) 2003 Gergely Nagy, Shawn Betts, Jay Belanger
;;
;; Authors: Gergely Nagy <algernon@debian.org>,
;;          Shawn Betts <sabetts@users.sourceforge.net>,
;;          Jay Belanger <belanger@truman.edu>,
;; Maintainer: Gergely Nagy <algernon@debian.org>
;; Version: 0.2
;; Keywords: faces, ratpoison, X

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 2 of
;; the License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public
;; License along with this program; if not, write to the Free
;; Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
;; MA 02111-1307 USA

;;; Commentary:
;;
;; This file provides a major mode for editing .ratpoisonrc files, and
;; functions to access ratpoison from within Emacs.

;;; History:
;; Version 0.2:
;;  - Added command-interpreter (from Shawn and Jay)
;;  - Added info-lookup functions (from Jay)
;;  - renamed to ratpoison.el
;;  - far better font-locking
;; Version 0.1:
;;  - initial version

;;; Todo:
;; - auto-completion of commands
;; - probably a bunch of other things

(require 'font-lock)
(require 'generic-x)

(defvar ratpoison-commands-0
  (list
   "abort"
   "banish"
   "clock"
   "curframe"
   "delete"
   "focus"
   "focusup"
   "focusdown"
   "focusleft"
   "focusright"
   "meta"
   "help"
   "info"
   "kill"
   "lastmsg"
   "redisplay"
   "restart"
   "next"
   "only"
   "other"
   "prev"
   "quit"
   "remove"
   "split"
   "hsplit"
   "version"
   "vsplit"
   ))

(defvar ratpoison-commands-rest
  (list
   "bind"
   "chdir"
   "colon"
   "defbarloc"
   "msgwait"
   "defborder"
   "deffont"
   "definputwidth"
   "defmaxsizepos"
   "defpadding"
   "deftranspos"
   "defwaitcursor"
   "defwinfmt"
   "defwinname"
   "defwinpos"
   "deffgcolor"
   "defbgcolor"
   "escape"
   "echo"
   "exec"
   "newwm"
   "number"
   "pos"
   "rudeness"
   "select"
   "setenv"
   "source"
   "startup_message"
   "title"
   "unbind"
   "unsetenv"
   "windows"
   ))

;; ratpoisonrc-mode
(define-generic-mode 'ratpoisonrc-mode
  ;; comments
  (list ?#)
  ;; keywords
  nil
  ;; font-lock stuff
  (list
    ;; commands without arguments
    (generic-make-keywords-list
     ratpoison-commands-0 font-lock-builtin-face "^[ \t]*")
    ;; commands with arguments
    (generic-make-keywords-list
     ratpoison-commands-rest font-lock-builtin-face "^[ \t]*" "[ \t]+")
    ;; exec <arg>
    (list "^[ \t]*\\(exec\\)[ \t]+\\(.*\\)"
          '(1 'font-lock-builtin-face)
          '(2 'font-lock-string-face))
    ;; arguments, the first is a keyword, the rest is tring
    (list (concat
                (car (generic-make-keywords-list
                      ratpoison-commands-rest font-lock-builtin-face "^[ \t]*" "[ \t]+"))
                "\\([0-9a-zA-Z\\/\\.\\-]+\\)[ \t]*\\(.*\\)")
          '(2 'font-lock-keyword-face)
          '(3 'font-lock-string-face)))
  ;; auto-mode alist
  (list "\\.ratpoisonrc\\'")
  ;; additional setup functions
  (list 'ratpoisonrc-mode-setup)
  "Generic mode for ratpoison configuration files.")

(defun ratpoisonrc-mode-setup()
  (defvar ratpoisonrc-mode-keymap (make-sparse-keymap)
    "Keymap for ratpoisonrc-mode")
  (define-key ratpoisonrc-mode-keymap "\C-c\C-e" 'ratpoison-line)
  (use-local-map ratpoisonrc-mode-keymap))

(provide 'ratpoisonrc-mode)

;; Ratpoison access
; Groups & Variables
(defgroup ratpoison nil "Ratpoison access"
  :group 'languages
  :prefix "ratpoison-")

(defcustom ratpoison-program "ratpoison"
  "The command to call the window manager."
  :group 'ratpoison
  :type 'string)

; Command stuff
(defun ratpoison-command (command)
  (interactive "sRP Command: ")
  (call-process ratpoison-program nil nil nil "-c" command))

(defun ratpoison-command-on-region (start end)
  (interactive "r")
  (mapcar 'ratpoison-command
          (split-string (buffer-substring start end)
                        "\n")))

(defun ratpoison-line ()
  "Send the current line to ratpoison."
  (interactive)
  (ratpoison-command
   (buffer-substring-no-properties
    (line-beginning-position)
    (line-end-position))))

;; Documentation
(defun ratpoison-manual ()
  "Call up the ratpoison info page."
  (interactive)
  (info "ratpoison"))

(defun ratpoison-manual-commands ()
  "Call up the info page listing the ratpoison commands."
  (interactive)
  (info "(ratpoison) Commands"))

(provide 'ratpoison)
